retry.go

```diff
@@ -0,0 +1,299 @@
+	defer rc.l.Unlock()
+
+	cons, err := rc.broker.Consumer(rc.queue)
+	if err != nil {
+		return nil, err
+	}
+	rc.cons = cons
+	return cons, nil
+}
+
+// discard is used to remove a broken consumer.
+func (rc *retryConsumer) discard(cons broker.Consumer) {
+	if cons == nil {
+		return
+	}
+	cons.Close()
+
+	rc.l.Lock()
+	defer rc.l.Unlock()
+	if cons == rc.cons {
+		rc.cons = nil
+	}
+}
+
+// Close closes the consumer.
+func (rc *retryConsumer) Close() error {
+	cons, err := rc.consumer(false)
+	if err != nil || cons == nil {
+		return err
+	}
+	return cons.Close()
+}
+
+// Ack marks message(s) as delivered.
+func (rc *retryConsumer) Ack() error {
+	cons, err := rc.consumer(false)
+	if err != nil || cons == nil {
+		return err
+	}
+	err = cons.Ack()
+	if err != nil {
+		log.Printf("[ERR] relay: consumer failed ack: %v", err)
+		rc.discard(cons)
+	}
+	return err
+}
+
+// Nack sends message(s) back to the queue.
+func (rc *retryConsumer) Nack() error {
+	cons, err := rc.consumer(false)
+	if err != nil || cons == nil {
+		return err
+	}
+	err = cons.Nack()
+	if err != nil {
+		log.Printf("[ERR] relay: consumer failed nack: %v", err)
+		rc.discard(cons)
+	}
+	return err
+}
+
+// ConsumeAck is used to consume with automatic ack. The consume operation
+// is able to be retried, but if a message is consumed and the acknowledgement
+// fails, the broker will re-deliver the message. Because of this, it is
+// critical that the returned error is checked, as it is possible that both
+// the output is written and an error is encountered if we see connection
+// loss between Consume() and Ack().
+func (rc *retryConsumer) ConsumeAck(out interface{}) error {
+	if err := rc.Consume(out); err != nil {
+		return err
+	}
+	return rc.Ack()
+}
+
+// Consume consumes a single message from the queue.
+func (rc *retryConsumer) Consume(out interface{}) error {
+	return rc.ConsumeTimeout(out, -1)
+}
+
+// ConsumeTimeout consumes a single message from the queue with an upper bound
+// on the time spent waiting. This behaves slightly different in the retry
+// broker, as we may encounter errors getting a valid consumer. In this case,
+// the timeout may actually be longer than specified to allow reconnection
+// attempts to take place.
+func (rc *retryConsumer) ConsumeTimeout(out interface{}, timeout time.Duration) error {
+	for i := 0; ; i++ {
+		cons, err := rc.consumer(true)
+		if err != nil {
+			goto RETRY
+		}
+
+		err = cons.ConsumeTimeout(out, timeout)
+		if err == nil {
+			break
+		}
+
+		// Respect the case where the consume legitimately times out.
+		if err == TimedOut {
+			return err
+		}
+
+	RETRY:
+		log.Printf("[ERR] relay: consumer got error: %v", err)
+		rc.discard(cons)
+		if i == rc.attempts {
+			log.Printf("[ERR] relay: consumer giving up after %d attempts", i)
+			return err
+		}
+		wait := rc.min * (1 << uint(i))
+		if wait > rc.max {
+			wait = rc.max
+		}
+		log.Printf("[DEBUG] relay: consumer retrying in %s", wait)
+		time.Sleep(wait)
```

Comment:
We should also downward adjust timeout when we wait, other wise the potential timeout added to the wait